README.md

# curve25519-dalek-fiat

[![curve25519-dalek-fiat on crates.io](https://img.shields.io/crates/v/curve25519-dalek-fiat)](https://crates.io/crates/curve25519-dalek-fiat)
[![Documentation (latest release)](https://docs.rs/curve25519-dalek-fiat/badge.svg)](https://docs.rs/curve25519-dalek-fiat/)

<img
 width="33%"
 align="right"
 src="https://doc.dalek.rs/assets/dalek-logo-clear.png"/>
 
# About

This is a thin fork of the [`curve25519-dalek`][curve25519-dalek] project, in order to expose a formally
verified backend supplied by the [`fiat-crypto`][fiat crypto] project, where 
primitive curve operations are extracted from Coq proofs of arithmetic correctness.

**A pure-Rust implementation of group operations on Ristretto and Curve25519.**

`curve25519-dalek` is a library providing group operations on the Edwards and
Montgomery forms of Curve25519, and on the prime-order Ristretto group.

`curve25519-dalek` is not intended to provide implementations of any particular
crypto protocol.  Rather, implementations of those protocols (such as
[`x25519-dalek`][x25519-dalek] and [`ed25519-dalek`][ed25519-dalek]) should use
`curve25519-dalek` as a library.

`curve25519-dalek` is intended to provide a clean and safe _mid-level_ API for use
implementing a wide range of ECC-based crypto protocols, such as key agreement,
signatures, anonymous credentials, rangeproofs, and zero-knowledge proof
systems.

In particular, `curve25519-dalek` implements Ristretto, which constructs a
prime-order group from a non-prime-order Edwards curve. This provides the
speed and safety benefits of Edwards curve arithmetic, without the pitfalls of
cofactor-related abstraction mismatches.

# Use

To import `curve25519-dalek-fiat`, add the following to the dependencies section of
your project's `Cargo.toml`:
```toml
curve25519-dalek-fiat = "0.1.0"
```

See `CHANGELOG.md` for more details.

# Backends and Features

The `nightly` feature enables features available only when using a Rust nightly
compiler.  In particular, it is required for rendering documentation and for
the SIMD backends.

Curve arithmetic is implemented using one of the following backends:

* a `fiat_u64_backend` to use a verified u64 backend supplied by the [`fiat-crypto`][fiat crypto] crate;
* a `u32` backend using serial formulas and `u64` products;
* a `u64` backend using serial formulas and `u128` products;
* an `avx2` backend using [parallel formulas][parallel_doc] and `avx2` instructions (sets speed records);
* an `ifma` backend using [parallel formulas][parallel_doc] and `ifma` instructions (sets speed records);

By default the `u64` backend is selected.  To select a specific backend, use:
```sh
cargo build --no-default-features --features "std fiat_u64_backend"
cargo build --no-default-features --features "std u32_backend"
cargo build --no-default-features --features "std u64_backend"
# Requires nightly, RUSTFLAGS="-C target_feature=+avx2" to use avx2
cargo build --no-default-features --features "std simd_backend"
# Requires nightly, RUSTFLAGS="-C target_feature=+avx512ifma" to use ifma
cargo build --no-default-features --features "std simd_backend"
```
Crates using `curve25519-dalek` can either select a backend on behalf of their
users, or expose feature flags that control the `curve25519-dalek` backend.

The `std` feature is enabled by default, but it can be disabled for no-`std`
builds using `--no-default-features`.  Note that this requires explicitly
selecting an arithmetic backend using one of the `_backend` features.
If no backend is selected, compilation will fail.

# Safety

The `curve25519-dalek` types are designed to make illegal states
unrepresentable.  For example, any instance of an `EdwardsPoint` is
guaranteed to hold a point on the Edwards curve, and any instance of a
`RistrettoPoint` is guaranteed to hold a valid point in the Ristretto
group.

All operations are implemented using constant-time logic (no
secret-dependent branches, no secret-dependent memory accesses),
unless specifically marked as being variable-time code.
We believe that our constant-time logic is lowered to constant-time
assembly, at least on `x86_64` targets.

As an additional guard against possible future compiler optimizations,
the `subtle` crate places an optimization barrier before every
conditional move or assignment.  More details can be found in [the
documentation for the `subtle` crate][subtle_doc].

Some functionality (e.g., multiscalar multiplication or batch
inversion) requires heap allocation for temporary buffers.  All
heap-allocated buffers of potentially secret data are explicitly
zeroed before release.

However, we do not attempt to zero stack data, for two reasons.
First, it's not possible to do so correctly: we don't have control
over stack allocations, so there's no way to know how much data to
wipe.  Second, because `curve25519-dalek` provides a mid-level API,
the correct place to start zeroing stack data is likely not at the
entrypoints of `curve25519-dalek` functions, but at the entrypoints of
functions in other crates.

The implementation is memory-safe, and contains no significant
`unsafe` code.  The SIMD backend uses `unsafe` internally to call SIMD
intrinsics.  These are marked `unsafe` only because invoking them on an
inappropriate CPU would cause `SIGILL`, but the entire backend is only
compiled with appropriate `target_feature`s, so this cannot occur.

# Performance

Benchmarks are run using [`criterion.rs`][criterion]:

```sh
cargo bench --no-default-features --features "std fiat_u64_backend"
cargo bench --no-default-features --features "std u32_backend"
cargo bench --no-default-features --features "std u64_backend"
# Uses avx2 or ifma only if compiled for an appropriate target.
export RUSTFLAGS="-C target_cpu=native"
cargo bench --no-default-features --features "std simd_backend"
```

Performance is a secondary goal behind correctness, safety, and
clarity, but we aim to be competitive with other implementations.
The new `fiat_u64` backend incurs a 8-15% slowdown compared to
the original `u64` backend, depending on the EdDSA operation.

| group | ed25519_fiat_u64_backend | ed25519_u64_backend |
| ----- | ------------------------ | ------------------- |
| Ed25519 batch signature verification/128 | 1.10&nbsp;&nbsp;&nbsp;3.0±0.01ms | 1.00&nbsp;&nbsp;&nbsp;2.7±0.01ms |
| Ed25519 batch signature verification/16  | 1.09&nbsp;&nbsp;&nbsp;411.7±1.28µs | 1.00&nbsp;&nbsp;&nbsp;377.8±0.92µs |
| Ed25519 batch signature verification/256 | 1.09&nbsp;&nbsp;&nbsp;5.4±0.01ms | 1.00&nbsp;&nbsp;&nbsp;4.9±0.01ms |
| Ed25519 batch signature verification/32  | 1.08&nbsp;&nbsp;&nbsp;779.3±4.87µs | 1.00&nbsp;&nbsp;&nbsp;723.9±3.21µs |
| Ed25519 batch signature verification/4   | 1.09&nbsp;&nbsp;&nbsp;137.9±0.75µs | 1.00&nbsp;&nbsp;&nbsp;127.0±0.30µs |
| Ed25519 batch signature verification/64  | 1.15&nbsp;&nbsp;&nbsp;1590.2±44.34µs | 1.00&nbsp;&nbsp;&nbsp;1385.2±6.80µs |
| Ed25519 batch signature verification/8   | 1.09&nbsp;&nbsp;&nbsp;229.0±0.92µs | 1.00&nbsp;&nbsp;&nbsp;210.2±0.63µs |
| Ed25519 batch signature verification/96  | 1.11&nbsp;&nbsp;&nbsp;2.4±0.08ms | 1.00&nbsp;&nbsp;&nbsp;2.2±0.01ms |
| Ed25519 keypair generation               | 1.07&nbsp;&nbsp;&nbsp;17.9±0.07µs | 1.00&nbsp;&nbsp;&nbsp;16.7±0.09µs |
| Ed25519 signature verification           | 1.11&nbsp;&nbsp;&nbsp;51.1±0.26µs | 1.00&nbsp;&nbsp;&nbsp;46.1±0.29µs |
| Ed25519 signing                          | 1.05&nbsp;&nbsp;&nbsp;18.9±0.09µs | 1.00&nbsp;&nbsp;&nbsp;18.0±0.07µs |
| Ed25519 signing w/ an expanded secret key| 1.11&nbsp;&nbsp;&nbsp;18.6±0.13µs | 1.00&nbsp;&nbsp;&nbsp;16.8±0.13µs |
| Ed25519 strict signature verification    | 1.06&nbsp;&nbsp;&nbsp;53.0±0.33µs | 1.00&nbsp;&nbsp;&nbsp;50.0±0.15µs |

[curve25519-dalek]: https://github.com/dalek-cryptography/curve25519-dalek
[ed25519-dalek]: https://github.com/dalek-cryptography/ed25519-dalek
[x25519-dalek]: https://github.com/dalek-cryptography/x25519-dalek
[contributing]: https://github.com/dalek-cryptography/curve25519-dalek/blob/master/CONTRIBUTING.md
[docs-external]: https://doc.dalek.rs/curve25519_dalek/
[docs-internal]: https://doc-internal.dalek.rs/curve25519_dalek/
[criterion]: https://github.com/japaric/criterion.rs
[parallel_doc]: https://doc-internal.dalek.rs/curve25519_dalek/backend/vector/avx2/index.html
[subtle_doc]: https://doc.dalek.rs/subtle/
[fiat crypto]: https://github.com/mit-plv/fiat-crypto


CHANGELOG.md
# Changelog

## 0.X series

### 0.1.0

* Initial release of curve25519-dalek-fiat, which is a fork of the
  curve25519-dalek (at v3.0.3) in order to add in a verified backend via the
  fiat-crypto project.
* Update the rand_core dependency to 0.6

## 0.1.1

* Update zeroize dependency to 1.4.2 due to RUSTSEC-2021-0115


benches/dalek_benchmarks.rs
#![allow(non_snake_case)]

extern crate rand;
use rand::rngs::OsRng;
use rand::thread_rng;

#[macro_use]
extern crate criterion;

use criterion::BatchSize;
use criterion::Criterion;

extern crate curve25519_dalek_fiat;

use curve25519_dalek_fiat::constants;
use curve25519_dalek_fiat::scalar::Scalar;

static BATCH_SIZES: [usize; 5] = [1, 2, 4, 8, 16];
static MULTISCALAR_SIZES: [usize; 13] = [1, 2, 4, 8, 16, 32, 64, 128, 256, 384, 512, 768, 1024];

mod edwards_benches {
    use super::*;

    use curve25519_dalek_fiat::edwards::EdwardsPoint;

    fn compress(c: &mut Criterion) {
        let B = &constants::ED25519_BASEPOINT_POINT;
        c.bench_function("EdwardsPoint compression", move |b| b.iter(|| B.compress()));
    }

    fn decompress(c: &mut Criterion) {
        let B_comp = &constants::ED25519_BASEPOINT_COMPRESSED;
        c.bench_function("EdwardsPoint decompression", move |b| {
            b.iter(|| B_comp.decompress().unwrap())
        });
    }

    fn consttime_fixed_base_scalar_mul(c: &mut Criterion) {
        let B = &constants::ED25519_BASEPOINT_TABLE;
        let s = Scalar::from(897987897u64).invert();
        c.bench_function("Constant-time fixed-base scalar mul", move |b| {
            b.iter(|| B * &s)
        });
    }

    fn consttime_variable_base_scalar_mul(c: &mut Criterion) {
        let B = &constants::ED25519_BASEPOINT_POINT;
        let s = Scalar::from(897987897u64).invert();
        c.bench_function("Constant-time variable-base scalar mul", move |b| {
            b.iter(|| B * s)
        });
    }

    fn vartime_double_base_scalar_mul(c: &mut Criterion) {
        c.bench_function("Variable-time aA+bB, A variable, B fixed", |bench| {
            let mut rng = thread_rng();
            let A = &Scalar::random(&mut rng) * &constants::ED25519_BASEPOINT_TABLE;
            bench.iter_batched(
                || (Scalar::random(&mut rng), Scalar::random(&mut rng)),
                |(a, b)| EdwardsPoint::vartime_double_scalar_mul_basepoint(&a, &A, &b),
                BatchSize::SmallInput,
            );
        });
    }

    criterion_group! {
        name = edwards_benches;
        config = Criterion::default();
        targets =
        compress,
        decompress,
        consttime_fixed_base_scalar_mul,
        consttime_variable_base_scalar_mul,
        vartime_double_base_scalar_mul,
    }
}

mod multiscalar_benches {
    use super::*;

    use curve25519_dalek_fiat::edwards::EdwardsPoint;
    use curve25519_dalek_fiat::edwards::VartimeEdwardsPrecomputation;
    use curve25519_dalek_fiat::traits::MultiscalarMul;
    use curve25519_dalek_fiat::traits::VartimeMultiscalarMul;
    use curve25519_dalek_fiat::traits::VartimePrecomputedMultiscalarMul;

    fn construct_scalars(n: usize) -> Vec<Scalar> {
        let mut rng = thread_rng();
        (0..n).map(|_| Scalar::random(&mut rng)).collect()
    }

    fn construct_points(n: usize) -> Vec<EdwardsPoint> {
        let mut rng = thread_rng();
        (0..n)
            .map(|_| &Scalar::random(&mut rng) * &constants::ED25519_BASEPOINT_TABLE)
            .collect()
    }

    fn construct(n: usize) -> (Vec<Scalar>, Vec<EdwardsPoint>) {
        (construct_scalars(n), construct_points(n))
    }

    fn consttime_multiscalar_mul(c: &mut Criterion) {
        c.bench_function_over_inputs(
            "Constant-time variable-base multiscalar multiplication",
            |b, &&size| {
                let points = construct_points(size);
                // This is supposed to be constant-time, but we might as well
                // rerandomize the scalars for every call just in case.
                b.iter_batched(
                    || construct_scalars(size),
                    |scalars| EdwardsPoint::multiscalar_mul(&scalars, &points),
                    BatchSize::SmallInput,
                );
            },
            &MULTISCALAR_SIZES,
        );
    }

    fn vartime_multiscalar_mul(c: &mut Criterion) {
        c.bench_function_over_inputs(
            "Variable-time variable-base multiscalar multiplication",
            |b, &&size| {
                let points = construct_points(size);
                // Rerandomize the scalars for every call to prevent
                // false timings from better caching (e.g., the CPU
                // cache lifts exactly the right table entries for the
                // benchmark into the highest cache levels).
                b.iter_batched(
                    || construct_scalars(size),
                    |scalars| EdwardsPoint::vartime_multiscalar_mul(&scalars, &points),
                    BatchSize::SmallInput,
                );
            },
            &MULTISCALAR_SIZES,
        );
    }

    fn vartime_precomputed_pure_static(c: &mut Criterion) {
        c.bench_function_over_inputs(
            "Variable-time fixed-base multiscalar multiplication",
            move |b, &&total_size| {
                let static_size = total_size;

                let static_points = construct_points(static_size);
                let precomp = VartimeEdwardsPrecomputation::new(&static_points);
                // Rerandomize the scalars for every call to prevent
                // false timings from better caching (e.g., the CPU
                // cache lifts exactly the right table entries for the
                // benchmark into the highest cache levels).
                b.iter_batched(
                    || construct_scalars(static_size),
                    |scalars| precomp.vartime_multiscalar_mul(&scalars),
                    BatchSize::SmallInput,
                );
            },
            &MULTISCALAR_SIZES,
        );
    }

    fn vartime_precomputed_helper(c: &mut Criterion, dynamic_fraction: f64) {
        let label = format!(
            "Variable-time mixed-base multiscalar multiplication ({:.0}pct dyn)",
            100.0 * dynamic_fraction,
        );
        c.bench_function_over_inputs(
            &label,
            move |b, &&total_size| {
                let dynamic_size = ((total_size as f64) * dynamic_fraction) as usize;
                let static_size = total_size - dynamic_size;

                let static_points = construct_points(static_size);
                let dynamic_points = construct_points(dynamic_size);
                let precomp = VartimeEdwardsPrecomputation::new(&static_points);
                // Rerandomize the scalars for every call to prevent
                // false timings from better caching (e.g., the CPU
                // cache lifts exactly the right table entries for the
                // benchmark into the highest cache levels).  Timings
                // should be independent of points so we don't
                // randomize them.
                b.iter_batched(
                    || {
                        (
                            construct_scalars(static_size),
                            construct_scalars(dynamic_size),
                        )
                    },
                    |(static_scalars, dynamic_scalars)| {
                        precomp.vartime_mixed_multiscalar_mul(
                            &static_scalars,
                            &dynamic_scalars,
                            &dynamic_points,
                        )
                    },
                    BatchSize::SmallInput,
                );
            },
            &MULTISCALAR_SIZES,
        );
    }

    fn vartime_precomputed_00_pct_dynamic(c: &mut Criterion) {
        vartime_precomputed_helper(c, 0.0);
    }

    fn vartime_precomputed_20_pct_dynamic(c: &mut Criterion) {
        vartime_precomputed_helper(c, 0.2);
    }

    fn vartime_precomputed_50_pct_dynamic(c: &mut Criterion) {
        vartime_precomputed_helper(c, 0.5);
    }

    criterion_group! {
        name = multiscalar_benches;
        // Lower the sample size to run the benchmarks faster
        config = Criterion::default().sample_size(15);
        targets =
        consttime_multiscalar_mul,
        vartime_multiscalar_mul,
        vartime_precomputed_pure_static,
        vartime_precomputed_00_pct_dynamic,
        vartime_precomputed_20_pct_dynamic,
        vartime_precomputed_50_pct_dynamic,
    }
}

mod ristretto_benches {
    use super::*;
    use curve25519_dalek_fiat::ristretto::RistrettoPoint;

    fn compress(c: &mut Criterion) {
        c.bench_function("RistrettoPoint compression", |b| {
            let B = &constants::RISTRETTO_BASEPOINT_POINT;
            b.iter(|| B.compress())
        });
    }

    fn decompress(c: &mut Criterion) {
        c.bench_function("RistrettoPoint decompression", |b| {
            let B_comp = &constants::RISTRETTO_BASEPOINT_COMPRESSED;
            b.iter(|| B_comp.decompress().unwrap())
        });
    }

    fn double_and_compress_batch(c: &mut Criterion) {
        c.bench_function_over_inputs(
            "Batch Ristretto double-and-encode",
            |b, &&size| {
                let mut rng = OsRng;
                let points: Vec<RistrettoPoint> = (0..size)
                    .map(|_| RistrettoPoint::random(&mut rng))
                    .collect();
                b.iter(|| RistrettoPoint::double_and_compress_batch(&points));
            },
            &BATCH_SIZES,
        );
    }

    criterion_group! {
        name = ristretto_benches;
        config = Criterion::default();
        targets =
        compress,
        decompress,
        double_and_compress_batch,
    }
}

mod montgomery_benches {
    use super::*;

    fn montgomery_ladder(c: &mut Criterion) {
        c.bench_function("Montgomery pseudomultiplication", |b| {
            let B = constants::X25519_BASEPOINT;
            let s = Scalar::from(897987897u64).invert();
            b.iter(|| B * s);
        });
    }

    criterion_group! {
        name = montgomery_benches;
        config = Criterion::default();
        targets = montgomery_ladder,
    }
}

mod scalar_benches {
    use super::*;

    fn scalar_inversion(c: &mut Criterion) {
        c.bench_function("Scalar inversion", |b| {
            let s = Scalar::from(897987897u64).invert();
            b.iter(|| s.invert());
        });
    }

    fn batch_scalar_inversion(c: &mut Criterion) {
        c.bench_function_over_inputs(
            "Batch scalar inversion",
            |b, &&size| {
                let mut rng = OsRng;
                let scalars: Vec<Scalar> = (0..size).map(|_| Scalar::random(&mut rng)).collect();
                b.iter(|| {
                    let mut s = scalars.clone();
                    Scalar::batch_invert(&mut s);
                });
            },
            &BATCH_SIZES,
        );
    }

    criterion_group! {
        name = scalar_benches;
        config = Criterion::default();
        targets =
        scalar_inversion,
        batch_scalar_inversion,
    }
}

criterion_main!(
    scalar_benches::scalar_benches,
    montgomery_benches::montgomery_benches,
    ristretto_benches::ristretto_benches,
    edwards_benches::edwards_benches,
    multiscalar_benches::multiscalar_benches,
);
